// @headlessui-float/vue/dist/headlessui-float.mjs
import { ref as s, unref as S, shallowRef as K, Fragment as ce, defineComponent as Q, toRef as Y, computed as se, watch as R, onMounted as de, onBeforeUnmount as me, provide as ge, cloneVNode as D, h as O, Transition as ye, createCommentVNode as ve, inject as be, nextTick as he, Teleport as we, mergeProps as je } from "vue";
import { arrow as xe, computePosition as Oe, offset as Se, flip as Fe, shift as Te, autoPlacement as Ae, hide as Ce, autoUpdate as Ee } from "@floating-ui/dom";
var z = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {}, Z = "Expected a function", q = 0 / 0, Ne = "[object Symbol]", pe = /^\s+|\s+$/g, Pe = /^[-+]0x[0-9a-f]+$/i, Be = /^0b[01]+$/i, $e = /^0o[0-7]+$/i, Ie = parseInt, Re = typeof z == "object" && z && z.Object === Object && z, We = typeof self == "object" && self && self.Object === Object && self, ze = Re || We || Function("return this")(), Me = Object.prototype, Ue = Me.toString, ke = Math.max, Le = Math.min, L = function() {
  return ze.Date.now();
};
function De(e, t, n) {
  var i, l, u, c, a, d, m = 0, h = !1, f = !1, w = !0;
  if (typeof e != "function")
    throw new TypeError(Z);
  t = J(t) || 0, M(n) && (h = !!n.leading, f = "maxWait" in n, u = f ? ke(J(n.maxWait) || 0, t) : u, w = "trailing" in n ? !!n.trailing : w);
  function T(r) {
    var g = i, y = l;
    return i = l = void 0, m = r, c = e.apply(y, g), c;
  }
  function P(r) {
    return m = r, a = setTimeout(j, t), h ? T(r) : c;
  }
  function W(r) {
    var g = r - d, y = r - m, x = t - g;
    return f ? Le(x, u - y) : x;
  }
  function A(r) {
    var g = r - d, y = r - m;
    return d === void 0 || g >= t || g < 0 || f && y >= u;
  }
  function j() {
    var r = L();
    if (A(r))
      return B(r);
    a = setTimeout(j, W(r));
  }
  function B(r) {
    return a = void 0, w && i ? T(r) : (i = l = void 0, c);
  }
  function U() {
    a !== void 0 && clearTimeout(a), m = 0, i = d = l = a = void 0;
  }
  function C() {
    return a === void 0 ? c : B(L());
  }
  function v() {
    var r = L(), g = A(r);
    if (i = arguments, l = this, d = r, g) {
      if (a === void 0)
        return P(d);
      if (f)
        return a = setTimeout(j, t), T(d);
    }
    return a === void 0 && (a = setTimeout(j, t)), c;
  }
  return v.cancel = U, v.flush = C, v;
}
function Ve(e, t, n) {
  var i = !0, l = !0;
  if (typeof e != "function")
    throw new TypeError(Z);
  return M(n) && (i = "leading" in n ? !!n.leading : i, l = "trailing" in n ? !!n.trailing : l), De(e, t, {
    leading: i,
    maxWait: t,
    trailing: l
  });
}
function M(e) {
  var t = typeof e;
  return !!e && (t == "object" || t == "function");
}
function _e(e) {
  return !!e && typeof e == "object";
}
function He(e) {
  return typeof e == "symbol" || _e(e) && Ue.call(e) == Ne;
}
function J(e) {
  if (typeof e == "number")
    return e;
  if (He(e))
    return q;
  if (M(e)) {
    var t = typeof e.valueOf == "function" ? e.valueOf() : e;
    e = M(t) ? t + "" : t;
  }
  if (typeof e != "string")
    return e === 0 ? e : +e;
  e = e.replace(pe, "");
  var n = Be.test(e);
  return n || $e.test(e) ? Ie(e.slice(2), n ? 2 : 8) : Pe.test(e) ? q : +e;
}
var Ge = Ve;
function F(e) {
  var n;
  if (e == null || e.value == null)
    return null;
  const t = (n = e.value.$el) != null ? n : e.value;
  return t.$el ? F(s(t)) : t;
}
function Xe(e = {}) {
  const t = s(null), n = s(null), i = s(void 0), l = s(void 0), u = s(S(e.placement || "bottom")), c = s(S(e.strategy || "absolute")), a = K({});
  return { x: i, y: l, placement: u, strategy: c, middlewareData: a, update: () => {
    const m = F(t), h = F(n);
    !m || !h || Oe(m, h, {
      placement: S(e.placement),
      strategy: S(e.strategy),
      middleware: S(e.middleware)
    }).then((f) => {
      i.value = f.x, l.value = f.y, u.value = f.placement, c.value = f.strategy, a.value = f.middlewareData;
    });
  }, reference: t, floating: n };
}
const Ye = (e) => {
  const { padding: t } = e;
  return {
    name: "arrow",
    options: e,
    fn(n) {
      const i = S(e.element);
      return i ? xe({ element: i, padding: t }).fn(n) : {};
    }
  };
};
function ee(e) {
  return e.reduce((t, n) => n.type === ce ? t.concat(ee(n.children)) : t.concat(n), []);
}
function V(e) {
  return e == null ? !1 : typeof e.type == "string" || typeof e.type == "object" || typeof e.type == "function";
}
function p(e) {
  return e = S(e), e && (e == null ? void 0 : e.nodeType) !== Node.COMMENT_NODE;
}
const rt = [
  "origin-bottom",
  "origin-top",
  "origin-right",
  "origin-left",
  "origin-bottom-left",
  "origin-bottom-right",
  "origin-top-left",
  "origin-top-right"
], qe = (e) => {
  switch (e) {
    case "top":
      return "origin-bottom";
    case "bottom":
      return "origin-top";
    case "left":
      return "origin-right";
    case "right":
      return "origin-left";
    case "top-start":
    case "right-end":
      return "origin-bottom-left";
    case "top-end":
    case "left-end":
      return "origin-bottom-right";
    case "right-start":
    case "bottom-start":
      return "origin-top-left";
    case "left-start":
    case "bottom-end":
      return "origin-top-right";
    default:
      return "";
  }
}, te = Symbol("ArrowState");
function ne(e) {
  const t = be(te, null);
  if (t === null) {
    const n = new Error(`<${e} /> must be in the <Float /> component.`);
    throw Error.captureStackTrace && Error.captureStackTrace(n, ne), n;
  }
  return t;
}
const Je = {
  as: {
    type: [String, Function, Object],
    default: "template"
  },
  floatingAs: {
    type: [String, Function, Object],
    default: "div"
  },
  show: {
    type: Boolean,
    default: null
  },
  placement: {
    type: String,
    default: "bottom-start"
  },
  strategy: {
    type: String,
    default: "absolute"
  },
  offset: [Number, Function, Object],
  shift: {
    type: [Boolean, Number, Object],
    default: !1
  },
  flip: {
    type: [Boolean, Number, Object],
    default: !1
  },
  arrow: {
    type: [Boolean, Number],
    default: !1
  },
  autoPlacement: {
    type: [Boolean, Object],
    default: !1
  },
  hide: {
    type: [Boolean, Object],
    default: !1
  },
  autoUpdate: {
    type: [Boolean, Object],
    default: !0
  },
  zIndex: {
    type: [Number, String],
    default: 9999
  },
  transitionName: String,
  transitionType: String,
  enter: String,
  enterFrom: String,
  enterTo: String,
  leave: String,
  leaveFrom: String,
  leaveTo: String,
  originClass: [String, Function],
  tailwindcssOriginClass: {
    type: Boolean,
    default: !1
  },
  portal: {
    type: [Boolean, String],
    default: !1
  },
  transform: {
    type: Boolean,
    default: !0
  },
  adaptiveWidth: {
    type: Boolean,
    default: !1
  },
  middleware: {
    type: [Array, Function],
    default: () => []
  }
}, Ke = Q({
  name: "Float",
  props: Je,
  emits: ["show", "hide", "update"],
  setup(e, { emit: t, slots: n, attrs: i }) {
    const l = s(!1), u = s(e.show !== null ? e.show : !1), c = Y(e, "placement"), a = Y(e, "strategy"), d = K(void 0), m = s(null), h = s(null), f = s(null), { x: w, y: T, placement: P, strategy: W, reference: A, floating: j, middlewareData: B, update: U } = Xe({
      placement: c,
      strategy: a,
      middleware: d
    }), C = se(() => typeof e.originClass == "function" ? e.originClass(P.value) : typeof e.originClass == "string" ? e.originClass : e.tailwindcssOriginClass ? qe(P.value) : ""), v = s(F(A)), r = s(F(j)), g = s(null);
    function y() {
      v.value = F(A), r.value = F(j);
    }
    function x() {
      p(v) && p(r) && (U(), t("update"));
    }
    R(c, () => {
      y(), x();
    }), R(a, () => {
      y(), x();
    }), R([
      () => e.offset,
      () => e.flip,
      () => e.shift,
      () => e.autoPlacement,
      () => e.arrow,
      () => e.hide,
      () => e.middleware
    ], () => {
      y();
      const o = [];
      (typeof e.offset == "number" || typeof e.offset == "object" || typeof e.offset == "function") && o.push(Se(e.offset)), (e.flip === !0 || typeof e.flip == "number" || typeof e.flip == "object") && o.push(Fe({
        padding: typeof e.flip == "number" ? e.flip : void 0,
        ...typeof e.flip == "object" ? e.flip : {}
      })), (e.shift === !0 || typeof e.shift == "number" || typeof e.shift == "object") && o.push(Te({
        padding: typeof e.shift == "number" ? e.shift : void 0,
        ...typeof e.shift == "object" ? e.shift : {}
      })), (e.autoPlacement === !0 || typeof e.autoPlacement == "object") && o.push(Ae(typeof e.autoPlacement == "object" ? e.autoPlacement : void 0)), (e.arrow === !0 || typeof e.arrow == "number") && o.push(Ye({
        element: m,
        padding: e.arrow === !0 ? 0 : e.arrow
      })), o.push(...typeof e.middleware == "function" ? e.middleware({
        referenceEl: v,
        floatingEl: r
      }) : e.middleware), (e.hide === !0 || typeof e.hide == "object") && o.push(Ce(typeof e.hide == "object" ? e.hide : void 0)), d.value = o, x();
    }, { immediate: !0 }), R(B, () => {
      var E, N;
      const o = B.value.arrow;
      h.value = (E = o == null ? void 0 : o.x) != null ? E : null, f.value = (N = o == null ? void 0 : o.y) != null ? N : null;
    });
    let $;
    function re() {
      p(v) && p(r) && e.autoUpdate !== !1 && ($ = Ee(v.value, r.value, Ge(x, 16), typeof e.autoUpdate == "object" ? e.autoUpdate : void 0));
    }
    function ie() {
      $ && ($(), $ = void 0);
    }
    let I;
    function ae() {
      y(), e.adaptiveWidth && typeof window < "u" && "ResizeObserver" in window && v.value && (I = new ResizeObserver(([o]) => {
        g.value = o.borderBoxSize.reduce((E, { inlineSize: N }) => E + N, 0);
      }), I.observe(v.value));
    }
    function oe() {
      I && (I.disconnect(), I = void 0, g.value = null);
    }
    async function _() {
      await he(), y(), p(v) && p(r) && u.value === !0 ? (re(), t("show")) : u.value === !1 && $ && (ie(), t("hide"));
    }
    return R(u, _), de(async () => {
      l.value = !0, ae(), await _();
    }), me(() => {
      oe();
    }), ge(te, {
      ref: m,
      placement: P,
      x: h,
      y: f
    }), () => {
      if (n.default) {
        let o = function(b) {
          return e.as === "template" ? b : typeof e.as == "string" ? O(e.as, i, b) : O(e.as, () => b);
        }, E = function(b) {
          return l.value && (e.portal === !0 || typeof e.portal == "string") ? O(we, {
            to: e.portal === !0 ? "body" : e.portal
          }, [b]) : b;
        }, N = function(b) {
          return e.floatingAs === "template" ? b : typeof e.floatingAs == "string" ? O(e.floatingAs, G, b) : O(e.floatingAs, () => b);
        };
        const [H, le] = ee(n.default() || []).filter(V);
        if (!V(H))
          return;
        const fe = {
          enterActiveClass: e.enter || C.value ? `${e.enter || ""} ${C.value}` : void 0,
          enterFromClass: e.enterFrom,
          enterToClass: e.enterTo,
          leaveActiveClass: e.leave || C.value ? `${e.leave || ""} ${C.value}` : void 0,
          leaveFromClass: e.leaveFrom,
          leaveToClass: e.leaveTo
        }, ue = {
          name: e.transitionName,
          type: e.transitionType,
          ...e.transitionName ? {} : fe,
          onBeforeEnter() {
            y(), u.value = !0;
          },
          onAfterLeave() {
            u.value = !1;
          }
        }, G = {
          ref: j,
          style: {
            ...e.transform ? {
              position: W.value,
              zIndex: e.zIndex,
              top: "0",
              left: "0",
              right: "auto",
              bottom: "auto",
              transform: `translate(${Math.round(w.value || 0)}px,${Math.round(T.value || 0)}px)`
            } : {
              position: W.value,
              zIndex: e.zIndex,
              top: `${T.value || 0}px`,
              left: `${w.value || 0}px`
            },
            width: e.adaptiveWidth && typeof g.value == "number" ? `${g.value}px` : ""
          }
        };
        return o([
          D(H, { ref: A }),
          E(N(O(ye, ue, () => {
            var X;
            const b = e.floatingAs === "template" ? G : null, k = D(le, b);
            return ((X = k.props) == null ? void 0 : X.unmount) === !1 ? (y(), x(), k) : typeof e.show != "boolean" || e.show ? k : ve();
          })))
        ]);
      }
    };
  }
}), Qe = {
  as: {
    type: [String, Object],
    default: "div"
  },
  offset: {
    type: Number,
    default: 4
  }
}, it = Q({
  name: "FloatArrow",
  props: Qe,
  setup(e, { slots: t, attrs: n }) {
    const { ref: i, placement: l, x: u, y: c } = ne("FloatArrow");
    return () => {
      var m;
      const a = {
        top: "bottom",
        right: "left",
        bottom: "top",
        left: "right"
      }[l.value.split("-")[0]], d = {
        left: typeof u.value == "number" ? `${u.value}px` : "",
        top: typeof c.value == "number" ? `${c.value}px` : "",
        right: "",
        bottom: "",
        [a]: `${e.offset * -1}px`
      };
      if (e.as === "template") {
        const h = { placement: l.value }, f = (m = t.default) == null ? void 0 : m.call(t, h), [w] = Array.isArray(f) ? f : [f];
        if (!w || !V(w))
          throw new Error("When the prop `as` of <FloatArrow /> is 'template', there must be contains 1 child element.");
        return D(w, { ref: i, style: d });
      }
      return O(e.as, Object.assign({}, n, { ref: i, style: d }));
    };
  }
});
function at(e) {
  return (t, { slots: n }) => O(Ke, je(e, t), n);
}
const Ze = [
  "Float",
  "FloatArrow"
];
function ot(e = {}) {
  const { prefix: t = "" } = e;
  return {
    type: "component",
    resolve: (n) => {
      if (n.startsWith(t)) {
        const i = n.substring(t.length);
        if (Ze.includes(i))
          return {
            name: i,
            from: "@headlessui-float/vue"
          };
      }
    }
  };
}
export {
  Ke as Float,
  it as FloatArrow,
  Qe as FloatArrowProps,
  Je as FloatProps,
  ot as HeadlessUiFloatResolver,
  Ye as arrow,
  at as createHighOrderFloatComponent,
  qe as tailwindcssOriginClassResolver,
  rt as tailwindcssOriginSafelist,
  Xe as useFloating
};
